[network] fix: ip address not retrieved if using cellular data

cruzach opened the pull request:
# Why

closes https://github.com/expo/expo/issues/7351

# How

If we aren't connected to wifi, the IP address returned on iOS is currently 0.0.0.0, and on android we would throw an error. At first I changed this to just return 0.0.0.0 so that the behavior would be consistent across both, but it makes sense to catch that case and fallback to getting the IP address from ipifiy.org (which is what we do in the web implementation already).

Related question- do we just want to rely on ipify for IP address on android & ios in the first place?

# Test Plan

Tested manually on android, IP address is returned both on data and wifi

cruzach committed:
return 0.0.0.0 if unable to get ip address (match ios)

cruzach committed:
cleanup some small things

cruzach committed:
fallback to ipify if unable to get ip natively

cruzach committed:
changelog and docs

cruzach committed:
fix lint

ide reviewed (changes requested):
This doesn't look like the correct fix. `InetAddress.getByAddress(new byte[] {0, 0, 0, 0})` runs fine. The issue seems to be that `BigInteger.valueOf(0).toByteArray()` produces an array that does not have four elements exactly.

cruzach committed:
return 0.0.0.0 if getByAddress throws

cruzach committed:
do not fallback to ipify on native

cruzach commented:
`getByAddress` will throw if the input isn't an array of length 4 or 16, so I changed it to just catch that error. Also removed the `ipify` fallback from native (it's still used on web)

ide commented on packages/expo-network/android/src/main/java/expo/modules/network/NetworkModule.java:
```diff
@@ -120,6 +120,19 @@ private NetworkStateType getNetworkCapabilities(NetworkCapabilities nc) {
+    byte[] ipByteArray = BigInteger.valueOf(ip).toByteArray();
+    try {
+      return InetAddress.getByAddress(ipByteArray).getHostAddress();
+    } catch (UnknownHostException e) {
```
This exception is thrown if the byte array is of the wrong length. While I think it's good to have this `catch` clause if we document this behavior in our API contract, this also means that IP addresses that start with a zero but aren't "0.0.0.0" (uncommon but valid within a network) will all get forced into this code path.

The byte array needs to be front-padded with zero bytes so that it always is four bytes long.

cruzach replied:
hm, I don't understand- why would an ip address that just starts with zero result in an array of the wrong length? 

I tried triggered the error by providing the `rawIpToString` method with the integer `16885760`, and got back `0.168.1.1` (as expected).

ide replied:
If it starts with a zero the integer will be too small to produce a four byte array. Try it.

cruzach replied:
ahh I see now, thanks for pointing that out

cruzach committed:
Update docs/pages/versions/unversioned/sdk/network.md

Co-authored-by: James Ide <ide@users.noreply.github.com>

cruzach committed:
ensure array passed to 'getByAddress' is 4 bytes long

cruzach committed:
use inputArray for consistent variable naming

ide reviewed (approved):
Make sure to test this with `0` (and maybe `1`) to make sure the padding logic works and that we don't take the try-catch code path.
